### PR TITLE
PP-843: pbsnode reports mem in b(bytes) instead of kb for cray login node.

### PR DESCRIPTION
**Pull request by dilip-krishnan**
#### Issue-ID

* **[PP-843](https://pbspro.atlassian.net/browse/PP-843)**

#### Problem description
* pbsnodes -av output shows resources_available.mem in bytes instead of kilobytes for
cray_login node

#### Cause / Analysis
* While creating inventory vnodes, cray_login node is re-initialise. Hence while setting memory resource
no unit was specified and pbs by default took the value to be in bytes instead of kilobytes
#### Solution description
* Specify the unit while setting the memory resource.

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [x] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: test/tests/functional/pbs_apbasil_inventory.py [hunk 1]
```diff
@@ -224,3 +224,40 @@ def test_InventoryVnodes(self):
...
+        self.mom.unset_mom_config('$alps_client', False)
+        self.reset_nodes(mom_id)
+
+        try:
+            pbs_node = self.server.status(NODE, id=mom_id)[0]
+
+        except PbsStatusError:
+            self.assertFalse(
+                "Cray login node %s doesn't exist on pbs server" % (mom_id))
+
```
Comment (riyazhakki) [inline test/tests/functional/pbs_apbasil_inventory.py]: Hi Dilip,
Can't we have Line 235 and 244 on same try block?

File: test/tests/functional/pbs_apbasil_inventory.py [hunk 2]
```diff
@@ -224,3 +224,40 @@ def test_InventoryVnodes(self):
...
+        for rsc, val in pbs_node:
+            if rsc in cray_login_node:
+                rval = cray_login_node[rsc]
+                if rval == val:
+                    self.logger.info(
+                        "%s\t: node has %s=%s" % (mom_id, rsc, rval))
+                    self.assertTrue(True)
```
Comment (riyazhakki) [inline test/tests/functional/pbs_apbasil_inventory.py]: Why is this line required?

File: test/tests/functional/pbs_apbasil_inventory.py (hunk 1 above)
Comment (dilip-krishnan) [reply test/tests/functional/pbs_apbasil_inventory.py]: Hi @riyazhakki , 
No we can't have them in same try block, because the call in line 244 is after calling reset_nodes() function, which deletes all the existing node and add the mom node again.

Regards
Dilip

File: test/tests/functional/pbs_apbasil_inventory.py (hunk 2 above)
Comment (dilip-krishnan) [reply test/tests/functional/pbs_apbasil_inventory.py]: This is just an indication of expected behavior.

File: test/tests/functional/pbs_apbasil_inventory.py (hunk 1 above)
Comment (riyazhakki) [reply test/tests/functional/pbs_apbasil_inventory.py]: I feel a try/except block can be avoided if we keep line 235,240,241 and 244 in one try block.
What say?

Comment (bhroam) [reply test/tests/functional/pbs_apbasil_inventory.py]: Just as a note, the correct usage of assertFalse() is self.assertFalse(<condition>, msg).  With your usage, the message will be considered true, so it will fail, but will not print the message.  If you want it to always fail,  you want self.assertFalse(True, <msg>) (or alternately use assertTrue(False), <msg>).

File: test/tests/functional/pbs_apbasil_inventory.py (hunk 2 above)
Comment (bhroam) [reply test/tests/functional/pbs_apbasil_inventory.py]: self.assertTrue(True) is a useless statement.  It will always pass.  If you want to say something to the test reader/runner, modify your log message.

File: test/tests/functional/pbs_apbasil_inventory.py [hunk 3]
```diff
@@ -224,3 +224,40 @@ def test_InventoryVnodes(self):
...
+            pbs_node = self.server.status(NODE, id=mom_id)[0]
+
+        except PbsStatusError:
+            self.assertFalse(
+                "Cray login node %s doesn't exist on pbs server" % (mom_id))
+
+        for rsc, val in pbs_node:
+            if rsc in cray_login_node:
+                rval = cray_login_node[rsc]
+                if rval == val:
```
Comment (bhroam) [inline test/tests/functional/pbs_apbasil_inventory.py]: Turn this if statement into an assertEquals().

File: test/tests/functional/pbs_apbasil_inventory.py (hunk 1 above)
Comment (bhroam) [reply test/tests/functional/pbs_apbasil_inventory.py]: I agree with Riyaz.  You can merge these two try/except blocks by putting all the code into one (including the removal of alps_client and reset_nodes().  You won't be able to really tell which of the two failed, but I don't think it is really necessary.

Comment (dilip-krishnan) [reply test/tests/functional/pbs_apbasil_inventory.py]: Hi @riyazhakki, @bhroam ,
    Differentiating which, out of the two node failed was the exact reason why two try blocks were needed. Since these nodes are results of two different set of code.One is created the default way of creating node, while cray login node is created through alps code. Also i noticed that I kept the log message same in both the places, which is a mistake.   

Unless you guys really want me to merge these try block, i would like to keep it with better log message.

Warm regards
Dilip

Comment (riyazhakki) [reply test/tests/functional/pbs_apbasil_inventory.py]: Hey @dilip-krishnan, merging these two try block looks ok to me unless you feel necessary to print the distinct messages for each exception.

**Comment by dilip-krishnan** [comment]
Hi @bhroam , @riyazhakki ,
    I have addressed your comments, please review the latest commit.

Regards
Dilip

File: test/tests/functional/pbs_apbasil_inventory.py [hunk 4]
```diff
@@ -233,31 +233,22 @@ def test_cray_login_node_memory_unit(self):
...
 
         except PbsStatusError:
-            self.assertFalse(
-                "Cray login node %s doesn't exist on pbs server" % (mom_id))
+            self.assertFalse(True,
+                             "Mom node %s doesn't exist on pbs server"
+                             % (mom_id))
 
         for rsc, val in pbs_node:
             if rsc in cray_login_node:
```
Comment (bhroam) [inline test/tests/functional/pbs_apbasil_inventory.py]: Instead of this if case, do a self.assertTrue(rsc in cray_login_node, msg).  If cases where the else case is an assertFalse(True) aren't the right approach.

Comment (dilip-krishnan) [reply test/tests/functional/pbs_apbasil_inventory.py]: Thanks for this comment, that does make sense and looks way better.

**Comment by riyazhakki** [comment]
Looks good.

**Comment by bhroam** [comment]
You have your sign offs.  Please squash and rebase for the merge.

**Comment by dilip-krishnan** [comment]
Hi @bhroam ,
    I have squashed and rebase my PR.

Regards
Dilip

**Comment by dilip-krishnan** [comment]
Hi @bhroam ,
    I have rebase and squashed this PR, please take it forward for merge.

Warm regards 
Dilip

**Comment by lisa-altair** [comment]
@dilip-krishnan I'm afraid my comment is coming a bit late, but someone asked me to look into why your test was failing...I did and so I'm commenting now.  As you determined it is because everything is compared and a  "licensing" attribute is missing when you did the reboot.   As you discovered, this test will fail with socket based licenses, but the test can be written so it works in all scenarios.
